[SPARK-43361][PROTOBUF] spark-protobuf: allow serde with enum as ints

justaparth opened the pull request:
https://issues.apache.org/jira/browse/SPARK-43361

### What changes were proposed in this pull request?
When deserializing protobuf enum fields, the spark-protobuf library will deserialize them as string values based on the enum name in the proto. E.g. 

```
message Person {
  enum Job {
    NOTHING = 0;
    ENGINEER = 1;
    DOCTOR = 2;
  }
  Job job = 1;
}
```
And we have a message like

```
Person(job=ENGINEER)
```

Then the deserialized value will be:

```
{"job": "ENGINEER"}
```

However it can be useful to deserialize the enum integer value rather than the name, namely:

```
{"job": 1}
```

This commit adds that functionality to spark protobuf via an option, "enums.as.ints". In addition, it also allows serializing such fields back into protobuf via `to_protobuf`.



Examples in other libraries:

- protobuf-java-util JsonFormat: https://javadoc.io/doc/com.google.protobuf/protobuf-java-util/3.10.0/com/google/protobuf/util/JsonFormat.Printer.html#printingEnumsAsInts--
- golang/protobuf jsonpb marshaler https://pkg.go.dev/github.com/golang/protobuf/jsonpb#Marshaler


### Why are the changes needed?
Adds new functionality that exists in other libraries.

### Does this PR introduce _any_ user-facing change?
Yes, adds a new option to from_protobuf.

### How was this patch tested?
Added unit tests confirming this behavior.

justaparth committed:
spark protobuf: allow serde with enum as ints

Adds a new option, enums.as.ints which allows deserializing enums as int
values. Symmetrically, we modify the serialization logic as well to allow
reading back structs where integer values are used for enums.

justaparth commented:
cc @rangadi @HyukjinKwon

rangadi commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/ProtobufDeserializer.scala:
```diff
@@ -260,6 +260,11 @@ private[sql] class ProtobufDeserializer(
 
+      case (ENUM, IntegerType) =>
+        (updater, ordinal, value) => {
+          updater.set(ordinal, protoType.getEnumType.findValueByName(value.toString).getNumber)
```
`value` itself would be of type `ProtocolMessageEnum` I think. Can do `value.asInstanceOf[ProtobufMessageEnum].getNumber()`. No need to convert to string then search.

justaparth replied:
ah thanks, i found it was of type `EnumDescriptor`, so i've updated this code + the code right above!

rangadi commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/ProtobufSerializer.scala:
```diff
@@ -110,6 +110,19 @@ private[sql] class ProtobufSerializer(
+        (getter, ordinal) =>
+          val data = getter.getInt(ordinal)
+          if (!enumValues.contains(data)) {
+            throw QueryCompilationErrors.cannotConvertCatalystTypeToProtobufEnumTypeError(
```
This is not type conversion error. Also it occurs at runtime, not at compile time. Check for other errors to see there is a better suited one.

justaparth replied:
yeah i agree, i copy/pasted the error from the existing logic above.

i've moved the error from `QueryCompilationErrors` to `QueryExecutionErrors` and slightly renamed it, wdyt? happy to change more or use a different error if that makes more sense though

rangadi reviewed (commented):
This is a useful option. Users should use this careful (can break compatibility down stream if they change it).

rangadi commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufOptions.scala:
```diff
@@ -141,6 +141,30 @@ private[sql] class ProtobufOptions(
+  // default deserialization will be:
+  // `{"job": "ENGINEER"}`
+  //
+  // But with this option set the deserialization will be:
```
Should we add a warning about users to be careful while changing it? It will change the schema and might be incompatible with downstream consumes.

rangadi commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufOptions.scala:
```diff
@@ -141,6 +141,30 @@ private[sql] class ProtobufOptions(
+  //
+  // But with this option set the deserialization will be:
+  // `{"job": 1}`
+  val enumsAsInts: Boolean =
```
Better to set this option for `jsonPrinter` in `ProtobufDeserializer` so that `Any` values deserialized as json strings will have ints too.

justaparth replied:
ah good catch! added + modified the test to assert it as well

justaparth committed:
update code to respond to comments; namely:
- add option to jsonformat.printer for any values
- move parsing error to queryexecutionerrors
- general code review comments

rangadi commented on connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufFunctionsSuite.scala:
```diff
@@ -361,15 +361,23 @@ class ProtobufFunctionsSuite extends QueryTest with SharedSparkSession with Prot
+            val fromProtoDF = df.select(
+              from_protobuf_wrapper($"value", name, descFilePathOpt, opts).as("value_from"))
+            val toProtoDF = fromProtoDF.select(
+              to_protobuf_wrapper($"value_from", name, descFilePathOpt).as("value_to"))
```
For `to_protobuf()` should we enforce `enum.as.ints` option is set while converting int to Enum?

justaparth replied:
yeah, i'm not sure either and also don't have a very strong opinion.

It felt safe to allow it by default because its more permissive, and also was hard for me to imagine how it could go wrong (if you have int values in your struct and you try to serialize and they're valid in the enum, then _probably_ thats what you were intending to do).

And this isn't _necessarily_ a good argument for what we should , but it appears that `JsonFormat`, for example, does this by default as well: https://github.com/protocolbuffers/protobuf/blob/main/java/util/src/main/java/com/google/protobuf/util/JsonFormat.java#L1920C33-L1946

So i'm inclined to leave it available by default, but yeah I don't really have a strong opinion either way!

rangadi replied:
Agree. Lets go with the current version (int will be used if the input schema is int).

rangadi reviewed (approved):
Thanks for the updates. LGTM. 
One comment about `to_protofbuf()` behavior if we should require this option to be set. I am not sure if should be strict with that.

HyukjinKwon commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufOptions.scala:
```diff
@@ -141,6 +141,33 @@ private[sql] class ProtobufOptions(
+  // Please note the output struct type will now contain an int column
+  // instead of string, so use caution if changing existing parsing logic.
+  val enumsAsInts: Boolean =
+    parameters.getOrElse("enums.as.ints", false.toString).toBoolean
```
Actually those options should better be documented at https://github.com/apache/spark/blob/master/docs/sql-data-sources-protobuf.md#deploying.

Since they are already undocumented there, I am fine without doing that in this PR though.

justaparth replied:
yeah thats a great point....

> Since they are already undocumented there, I am fine without doing that in this PR though.

yeah, maybe it makes sense in a new PR (which i'm happy to do) to add documentation related to all of the options available? since like you said, none of them seem to be documented

justaparth replied:
@HyukjinKwon went ahead and made a ticket for this https://issues.apache.org/jira/browse/SPARK-43409 !

HyukjinKwon commented:
@rangadi is it good to go?

rangadi commented:
@HyukjinKwon Yep, this looks good to go.

amaliujia commented:
Just wondering why this is relevant to Spark Connect (seeing `[CONNECT]` in the PR title)?

HyukjinKwon commented:
Merged to master.

justaparth commented:
> Just wondering why this is relevant to Spark Connect (seeing `[CONNECT]` in the PR title)?

ah sorry about that. i think i confused that this is under the `connector/` library with `CONNECT` which i saw in some other PR titles. In the future i'll add only `PROTOBUF` 👍

rangadi commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:
```diff
@@ -2812,4 +2813,18 @@ private[sql] object QueryExecutionErrors extends QueryErrorsBase {
+      data: String,
+      enumString: String): Throwable = {
+    new AnalysisException(
+      errorClass = "CANNOT_CONVERT_SQL_VALUE_TO_PROTOBUF_ENUM_TYPE",
```
@justaparth we forgot to update core/src/main/resources/error/error-classes.json and docs/sql-error-conditions.md in this PR.
Don't know how this got merged without failing. 
Could you send a follow up PR updating those files? Also, it will be better if we verify this error with `checkError()` in unit tests (see similar examples).
